R/create_model_out_submit_tmpl.R

```diff
@@ -0,0 +1,133 @@
+#' are excluded from the output. To include them as columns of `NA`s, use
+#' `include_opt_cols = FALSE`.
+#' @export
+#'
+#' @examples
+#' hub_con <- connect_hub(system.file("testhubs/flusight", package = "hubUtils"))
+#' create_model_out_submit_tmpl(hub_con)
+#' create_model_out_submit_tmpl(hub_con, required_only = TRUE)
+#' create_model_out_submit_tmpl(hub_con,
+#'   required_only = TRUE,
+#'   include_opt_cols = TRUE
+#' )
+#' # Specifying a round in a hub with multiple rounds
+#' hub_con <- connect_hub(system.file("testhubs/simple", package = "hubUtils"))
+#' create_model_out_submit_tmpl(hub_con, round_id = "2022-10-01")
+#' create_model_out_submit_tmpl(hub_con, round_id = "2022-10-29")
+#' create_model_out_submit_tmpl(hub_con,
+#'   round_id = "2022-10-29",
+#'   required_only = TRUE
+#' )
+#' create_model_out_submit_tmpl(hub_con,
+#'   round_id = "2022-10-29",
+#'   required_only = TRUE,
+#'   include_opt_cols = TRUE
+#' )
+create_model_out_submit_tmpl <- function(hub_con, config_tasks,
+                                         round_id = NULL,
+                                         required_only = FALSE,
+                                         include_opt_cols = FALSE) {
+  switch(rlang::check_exclusive(hub_con, config_tasks),
+    hub_con = {
+      checkmate::assert_class(hub_con, classes = "hub_connection")
+      config_tasks <- attr(hub_con, "config_tasks")
+    },
+    config_tasks = checkmate::assert_list(config_tasks)
+  )
+
+  round_task_ids <- get_round_task_id_names(
+    config_tasks,
+    round_id
+  )
+
+  tmpl_df <- expand_model_out_val_grid(config_tasks,
+    round_id = round_id,
+    required_only = required_only
+  )
+
+  opt_task_ids <- round_task_ids[!round_task_ids %in% names(tmpl_df)]
+
+  if (required_only && length(opt_task_ids) > 0L) {
+    n_mt <- n_model_tasks(config_tasks, round_id)
+    message_opt_tasks(opt_task_ids, n_mt, include_opt_cols)
+
+    if (include_opt_cols) {
+      conv_opt_task_ids <- create_hub_schema(
+        config_tasks,
+        partitions = NULL,
+        r_schema = TRUE
+      )[opt_task_ids]
+
+      tmpl_df[, opt_task_ids] <- NA
+      tmpl_df[, names(conv_opt_task_ids)] <- purrr::map2(
+        .x = names(conv_opt_task_ids),
+        .y = conv_opt_task_ids,
+        ~ get(paste0("as.", .y))(tmpl_df[[.x]])
+      )
+
+      tmpl_df <- tmpl_df[, c(round_task_ids, "output_type", "output_type_id")]
+    }
+  }
+  tmpl_df
+}
+
+get_round_model_tasks <- function(config_tasks, round_id) {
+  round_idx <- get_round_idx(
+    config_tasks,
+    round_id
+  )
+  purrr::pluck(
+    config_tasks,
+    "rounds",
+    round_idx,
+    "model_tasks"
+  )
+}
+
+n_model_tasks <- function(config_tasks, round_id) {
+  length(get_round_model_tasks(config_tasks, round_id))
+}
+
+
+get_round_task_id_names <- function(config_tasks, round_id) {
+  get_round_model_tasks(config_tasks, round_id) %>%
+    purrr::map(~ .x[["task_ids"]] %>%
+      names()) %>%
```

Comment:
minor question: should we be standardizing to use base R pipe instead of magrittr pipe?

Reply:
We agreed at the beginning of the project to use the magrittr pipe so as not to lock in users to R versions 4.1 and above. Happy to revisit this at some point though.

Reply:
Thanks for the reminder! Sounds good to me.